tfhe/src/core_crypto/gpu/algorithms/test/glwe_dot_product_with_clear.rs

```diff
@@ -0,0 +1,356 @@
+) -> Vec<Scalar>
+where
+    Scalar: UnsignedTorus,
+    InputCont: Container<Element = Scalar>,
+    KeyCont: Container<Element = Scalar>,
+{
+    let mut decrypted = PlaintextList::new(Scalar::ZERO, PlaintextCount(glwe.polynomial_size().0));
+    decrypt_glwe_ciphertext(glwe_secret_key, glwe, &mut decrypted);
+
+    let mut decoded = vec![Scalar::ZERO; decrypted.plaintext_count().0];
+    let ciphertext_modulus = glwe.ciphertext_modulus();
+    let delta = encryption_delta(bits_reserved_for_computation, ciphertext_modulus);
+
+    let decomposer = SignedDecomposer::new(
+        DecompositionBaseLog(
+            bits_reserved_for_computation
+                + ciphertext_modulus
+                    .get_power_of_two_scaling_to_native_torus()
+                    .ilog2() as usize,
+        ),
+        DecompositionLevelCount(1),
+    );
+
+    let decryption_modulus = Scalar::ONE << bits_reserved_for_computation;
+
+    for (decoded_value, decrypted_value) in decoded.iter_mut().zip(decrypted.iter()) {
+        *decoded_value =
+            (decomposer.closest_representable(*decrypted_value.0) / delta) % decryption_modulus;
+
+        if *decoded_value >= decryption_modulus / Scalar::TWO {
+            *decoded_value = (decryption_modulus - *decoded_value).wrapping_neg();
+        }
+    }
+    decoded
+}
+
+pub fn decrypt_lwe<Scalar, InputCont, KeyCont>(
+    lwe_secret_key: &LweSecretKey<KeyCont>,
+    lwe: &LweCiphertext<InputCont>,
+    bits_reserved_for_computation: usize,
+) -> Scalar
+where
+    Scalar: UnsignedTorus,
+    InputCont: Container<Element = Scalar>,
+    KeyCont: Container<Element = Scalar>,
+{
+    let decrypted_value = decrypt_lwe_ciphertext(lwe_secret_key, lwe);
+
+    let ciphertext_modulus = lwe.ciphertext_modulus();
+    let delta = encryption_delta(bits_reserved_for_computation, ciphertext_modulus);
+
+    let decomposer = SignedDecomposer::new(
+        DecompositionBaseLog(
+            bits_reserved_for_computation
+                + ciphertext_modulus
+                    .get_power_of_two_scaling_to_native_torus()
+                    .ilog2() as usize,
+        ),
+        DecompositionLevelCount(1),
+    );
+
+    let decryption_modulus = Scalar::ONE << bits_reserved_for_computation;
+
+    let mut decoded_value =
+        (decomposer.closest_representable(decrypted_value.0) / delta) % decryption_modulus;
+
+    if decoded_value >= decryption_modulus / Scalar::TWO {
+        decoded_value = (decryption_modulus - decoded_value).wrapping_neg();
+    }
+
+    decoded_value
+}
+/// Test that a GLWE can be multiplied with a list of clear polynomials.
+/// This test does not add encryption noise to ensure deterministic
+/// results. The polynomials are stored in a single CudaVec on device.
+fn glwe_dot_product_with_clear<Scalar: UnsignedTorus + CastFrom<usize>>(
+    _params: ClassicTestParams<Scalar>,
+) {
+    let mut rng = rand::thread_rng();
+
+    let poly_size = 2 << rng.gen_range(8usize..12);
```

Comment:
sometimes check with n_polys=poly_size, sometimes check with arbitrary number of polys